parcels/fieldset.py

```diff
@@ -772,41 +773,80 @@ def computeTimeChunk(self, time, dt):
 
-        # load in new data
         for f in self.get_fields():
             if type(f) in [VectorField, NestedField, SummedField] or not f.grid.defer_load or f.is_gradient or f.dataFiles is None:
                 continue
             g = f.grid
             if g.update_status == 'first_updated':  # First load of data
-                data = np.empty((g.tdim, g.zdim, g.ydim-2*g.meridional_halo, g.xdim-2*g.zonal_halo), dtype=np.float32)
-                f.loaded_time_indices = range(3)
-                for tind in f.loaded_time_indices:
-                    f.computeTimeChunk(data, tind)
+                data = da.empty((g.tdim, g.zdim, g.ydim-2*g.meridional_halo, g.xdim-2*g.zonal_halo), dtype=np.float32)
+                loaded_time_indices = range(3)
+                for tind in loaded_time_indices:
+                    data = f.computeTimeChunk(data, tind)
+                # ### do built-in computations on data
+                if f._scaling_factor:
+                    data *= f._scaling_factor
+                data[np.isnan(data)] = 0
+                if f.vmin is not None:
+                    data[data < f.vmin] = 0
+                if f.vmax is not None:
+                    data[data > f.vmax] = 0
+                if f.gradientx is not None:
+                    assert False
+                    f.gradient(update=True, tindex=tind)
                 f.data = f.reshape(data)
+                if not f.chunk_set:
+                    f.chunk_setup()
+                if len(f.load_chunk) > 0:
+                    f.load_chunk = np.where(f.load_chunk == 2, 1, f.load_chunk)
+                    f.load_chunk = np.where(f.load_chunk == 3, 0, f.load_chunk)
             elif g.update_status == 'updated':
-                data = np.empty((g.tdim, g.zdim, g.ydim-2*g.meridional_halo, g.xdim-2*g.zonal_halo), dtype=np.float32)
+                data = da.empty((g.tdim, g.zdim, g.ydim-2*g.meridional_halo, g.xdim-2*g.zonal_halo), dtype=np.float32)
                 if signdt >= 0:
-                    f.data[:2, :] = f.data[1:, :]
-                    f.loaded_time_indices = [2]
+                    f.filebuffers[0].dataset.close()
+                    f.filebuffers[:2] = f.filebuffers[1:]
+                    data = f.computeTimeChunk(data, 2)
                 else:
-                    f.data[1:, :] = f.data[:2, :]
-                    f.loaded_time_indices = [0]
-                f.computeTimeChunk(data, f.loaded_time_indices[0])
-                f.data[f.loaded_time_indices[0], :] = f.reshape(data)[f.loaded_time_indices[0], :]
-            else:
-                f.loaded_time_indices = []
-
-            # do built-in computations on data
-            for tind in f.loaded_time_indices:
+                    f.filebuffers[2].dataset.close()
+                    f.filebuffers[1:] = f.filebuffers[:2]
+                    data = f.computeTimeChunk(data, 0)
+                # do built-in computations on data
                 if f._scaling_factor:
```

Comment:
Some of the code below is copied from above. Put into a function so that we don't have repetition?